enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+---
+title: machine-health-checking
+authors:
+  - @bison
+  - @ingvagabund
+  - @enxebre
+  - @cynepco3hahue
+reviewers:
+  - @derekwaynecarr
+  - @michaelgugino
+  - @beekhof
+  - @mrunalp
+approvers:
+  - @enxebre
+  - @bison
+  - @derekwaynecarr
+  - @michaelgugino
+
+creation-date: 2019-09-09
+last-updated: 2019-09-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Machine health checking
+
+## Release Signoff Checklist
+
+ - [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+Enable opt in for automated health checking and remediation of unhealhty nodes backed by machines.
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
```

Comment:
Is there a reason out of the box not to have “reboot” be an option and just have the controller report if reboot is not possible for that provider?

I expect reboot to “just work” on cloud

Reply:
I would love reboot to be the default, especially with etcd on masters still being a snowflake.
The challenge as I understand it, is that the Machine API does not expose a platform independent reboot capability

Reply:
There is no reboot action in the machine-api. It would have to be implemented in the controller for each provider. The initial use-case driving this was bare metal, which is more complex. We'd like to leave this complexity out of the main health check controller.